File: app/src/main/java/com/ch/yoon/kakao/pay/imagesearch/ui/imagedetail/ImageDetailActivity.java
package com.ch.yoon.kakao.pay.imagesearch.ui.imagedetail;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.MenuItem;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModelProviders;

import com.ch.yoon.kakao.pay.imagesearch.R;
import com.ch.yoon.kakao.pay.imagesearch.databinding.ActivityImageDetailBinding;
import com.ch.yoon.kakao.pay.imagesearch.data.model.imagesearch.response.ImageDocument;
import com.ch.yoon.kakao.pay.imagesearch.ui.base.BaseActivity;

import java.util.Optional;

public class ImageDetailActivity extends BaseActivity<ActivityImageDetailBinding> {

    public static final String EXTRA_IMAGE_DOCUMENT_KEY = "EXTRA_IMAGE_DOCUMENT_KEY";

    private ActivityImageDetailBinding binding;

    public static Intent getImageDetailActivityIntent(@NonNull Context context,
                                                      @NonNull ImageDocument imageDocument) {
        final Intent imageDetailIntent = new Intent(context, ImageDetailActivity.class);
        imageDetailIntent.putExtra(ImageDetailActivity.EXTRA_IMAGE_DOCUMENT_KEY, imageDocument);
        return imageDetailIntent;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = binding(R.layout.activity_image_detail);

        final ImageDocument passedImageDocument = getIntent().getParcelableExtra(EXTRA_IMAGE_DOCUMENT_KEY);

        initBackArrow();
        initImageDetailViewModel(passedImageDocument);
        observeImageDetailViewModel();
    }

    @Override
    public boolean onOptionsItemSelected(final MenuItem item) {
        final int itemId = item.getItemId();
        if(itemId == android.R.id.home) {
            onBackPressed();
            return true;
        } else {
            return super.onOptionsItemSelected(item);
        }
    }

    private void initBackArrow() {
        setSupportActionBar(binding.mainToolbar);
        Optional.ofNullable(getSupportActionBar())
            .ifPresent(actionBar -> actionBar.setDisplayHomeAsUpEnabled(true));
    }

    private void initImageDetailViewModel(ImageDocument imageDocument) {
        final ImageDetailViewModel viewModel = ViewModelProviders.of(this, new ImageDetailViewModelFactory(
            getApplication()
        )).get(ImageDetailViewModel.class);

        viewModel.showImageDetailInfo(imageDocument);

        binding.setImageDetailViewModel(viewModel);
    }

    private void observeImageDetailViewModel() {
        binding.getImageDetailViewModel()
            .observeShowMessage()
            .observe(this, this::showToast);

        binding.getImageDetailViewModel()
            .observeMoveWebEvent()
            .observe(this, url -> {
                Uri webUri = Uri.parse(url);
                Intent movieWebIntent = new Intent(Intent.ACTION_VIEW, webUri);
                startActivity(movieWebIntent);
            });

        binding.getImageDetailViewModel()
            .observeFinishEvent()
            .observe(this, aVoid -> finish());
    }

    @Override
    public void onBackPressed() {
        binding.getImageDetailViewModel().onClickBackPress();
    }

}

File: app/src/main/java/com/ch/yoon/kakao/pay/imagesearch/ui/imagedetail/ImageDetailActivity.kt
package com.ch.yoon.kakao.pay.imagesearch.ui.imagedetail

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.MenuItem
import androidx.lifecycle.Observer
import com.ch.yoon.kakao.pay.imagesearch.R
import com.ch.yoon.kakao.pay.imagesearch.data.source.remote.kakao.response.KakaoImageDocument
import com.ch.yoon.kakao.pay.imagesearch.databinding.ActivityImageDetailBinding
import com.ch.yoon.kakao.pay.imagesearch.ui.base.BaseActivity
import org.koin.androidx.viewmodel.ext.android.viewModel

/**
 * Creator : ch-yoon
 * Date : 2019-10-29.
 */
class ImageDetailActivity : BaseActivity<ActivityImageDetailBinding>() {

    companion object {
        const val EXTRA_IMAGE_DOCUMENT_KEY = "EXTRA_IMAGE_DOCUMENT_KEY"

        fun getImageDetailActivityIntent(context: Context, kakaoImageDocument: KakaoImageDocument): Intent {
            return Intent(context, ImageDetailActivity::class.java).apply {
                putExtra(EXTRA_IMAGE_DOCUMENT_KEY, kakaoImageDocument)
            }
        }
    }

    private val imageDetailViewModel: ImageDetailViewModel by viewModel()

    override fun getLayoutId(): Int {
        return R.layout.activity_image_detail
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val passedImageDocument = intent.getParcelableExtra<KakaoImageDocument>(EXTRA_IMAGE_DOCUMENT_KEY)

        initBackArrow()
        initImageDetailViewModel(passedImageDocument)
        observeImageDetailViewModel()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val itemId = item.itemId
        return if (itemId == android.R.id.home) {
            onBackPressed()
            true
        } else {
            super.onOptionsItemSelected(item)
        }
    }

    private fun initBackArrow() {
        setSupportActionBar(binding.mainToolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
    }

    private fun initImageDetailViewModel(kakaoImageDocument: KakaoImageDocument) {
        binding.imageDetailViewModel = imageDetailViewModel
        if(isActivityFirstCreate) {
            imageDetailViewModel.showImageDetailInfo(kakaoImageDocument)
        }
    }

    private fun observeImageDetailViewModel() {
        val owner = this
        with(imageDetailViewModel) {
            showMessageEvent.observe(owner, Observer { message ->
                showToast(message)
            })

            moveToWebEvent.observe(owner, Observer { url ->
                val movieWebIntent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
                startActivity(movieWebIntent)
            })

            finishEvent.observe(owner, Observer {
                finish()
            })
        }
    }

    override fun onBackPressed() {
        imageDetailViewModel.onClickBackPress()
    }
}